src/OrchardCore.Modules/OrchardCore.Media/SecureMediaPermissions.cs

```diff
@@ -0,0 +1,155 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using OrchardCore.Media.Services;
+using OrchardCore.Security.Permissions;
+
+namespace OrchardCore.Media
+{
+    public class SecureMediaPermissions : IPermissionProvider
+    {
+        // Note: The ManageMediaFolder permission grants all access, so viewing must be implied by it too.
+        public static readonly Permission ViewMedia = new("ViewMediaContent", "View media content in all folders", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewRootMedia = new("ViewRootMediaContent", "View media content in the root folder", new[] { ViewMedia });
+        public static readonly Permission ViewOthersMedia = new("ViewOthersMediaContent", "View others media content", new[] { Permissions.ManageMediaFolder });
+        public static readonly Permission ViewOwnMedia = new("ViewOwnMediaContent", "View own media content", new[] { ViewOthersMedia });
+
+        private static readonly Permission s_viewMediaTemplate = new("ViewMediaContent_{0}", "View media content in {0}", new[] { ViewMedia });
+
+        private static Dictionary<ValueTuple<string, string>, Permission> s_permissionsByFolder = new();
+        private static readonly char[] s_trimSecurePathChars = ['/', '\\', ' '];
+        public static readonly ReadOnlyDictionary<string, Permission> PermissionTemplates = new(new Dictionary<string, Permission>()
+        {
+            { ViewMedia.Name, s_viewMediaTemplate },
+        });
+
+        private readonly MediaOptions _mediaOptions;
+        private readonly AttachedMediaFieldFileService _attachedMediaFieldFileService;
+        private readonly IMediaFileStore _fileStore;
+
+        public SecureMediaPermissions(
+            IOptions<MediaOptions> options,
+            IMediaFileStore fileStore,
+            AttachedMediaFieldFileService attachedMediaFieldFileService)
+        {
+            _mediaOptions = options.Value;
+            _fileStore = fileStore;
+            _attachedMediaFieldFileService = attachedMediaFieldFileService;
+        }
+
+        public async Task<IEnumerable<Permission>> GetPermissionsAsync()
+        {
+            // The ViewRootMedia permission must be implied by any sub folder permission.
+            var viewRootImpliedBy = new List<Permission>(ViewRootMedia.ImpliedBy);
+
+            var result = new List<Permission>()
+            {
+                ViewMedia,
+                new (ViewRootMedia.Name, ViewRootMedia.Description, viewRootImpliedBy),
+                ViewOthersMedia,
+                ViewOwnMedia
+            };
+
+            await foreach (var entry in _fileStore.GetDirectoryContentAsync())
```

Comment:
Yes, I was worried that there would be too many permissions if all subfolders would be included. Only the first level of folders are securable with this PR.

Reply:
How about unlocking that with an `IOptions` option? (See e.g. `MediaOptions`.)

Reply:
That would be a good addition, but do you mind if we add that later?

Reply:
I'd add it now. We at Lombiq implemented authenticated Media/file storage in unique ways for three projects, and all three of these needed permissions for subfolders. Now, that's not necessarily representative, but as far as real-life use cases go, it's significant, since all of these were pretty large projects.

Reply:
As much as I'd love to do that, I think I've to skip this addition for now. It takes too much time for me, as it would require a redesign of the authorization handler, e.g. on how to inherit permissions of subfolders. I'll reconsider that later.

Reply:
How about saying that if the folder has permissions then only that applies (i.e. no merging of permissions), otherwise the closest parent that has permissions. This is what we've done for the most recent project.

Reply:
I've also asked the people under the linked issues, in case everybody else would be happy without this: https://github.com/OrchardCMS/OrchardCore/issues/9369#issuecomment-2039754807

Reply:
Let's leave this for now